File: week_9/Final_Code/3.0 实时显示/Edge_Detection.py
import cv2
import numpy as np

def preprocess_image(image):
    
    Denoised_image = cv2.fastNlMeansDenoisingColored(image, None, 10, 10, 7, 21)
    Enhanced_image = cv2.convertScaleAbs(Denoised_image, alpha=1, beta=30)
    canny_image = cv2.Canny(Enhanced_image, 50, 150)
    
    return canny_image

def detect_quadrilaterals(image):
    contours, _ = cv2.findContours(image, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    quadrilaterals = []
    for contour in contours:
        area = cv2.contourArea(contour)
        if area >2000:
            peri = cv2.arcLength(contour, True)
            approx = cv2.approxPolyDP(contour, 0.01 * peri, True)

            if len(approx) == 4:
                quadrilaterals.append(approx)

    return quadrilaterals
def draw_points(image, quadrilaterals):  
  
    annotated_image = image.copy()  
      
    for quad in quadrilaterals:  
        cv2.drawContours(annotated_image, [quad], -1, (255, 0, 0), 2)  
  
        for i, point in enumerate(quad):  
            x, y = point[0]  
            cv2.circle(annotated_image, (x, y), 5, (0, 0, 255), -1)  
            cv2.putText(annotated_image, f'({x}, {y})', (x, y), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)  
            print(f"角点{i+1} 坐标: ({x}, {y})")  
              
        # 标记四边形的中心点  
        M = cv2.moments(quad)  
        if M['m00'] != 0: 
            cx = int(M['m10'] / M['m00'])  
            cy = int(M['m01'] / M['m00'])  
            cv2.circle(annotated_image, (cx, cy), 5, (0, 0, 255), -1) 
            cv2.putText(annotated_image, f'({cx}, {cy})', (cx, cy), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)  
            print(f"中点 坐标: ({cx}, {cy})")  
  
    return annotated_image  

# 主函数
if __name__ == "__main__":
    image = cv2.imread('week_9\src\img_5.jpeg')

    processed_image = preprocess_image(image)
    quadrilaterals = detect_quadrilaterals(processed_image)
    Annotated_Image = draw_points(image, quadrilaterals)

    cv2.namedWindow('Annotated Image', cv2.WINDOW_NORMAL)
    cv2.imshow('Annotated Image', Annotated_Image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: Edge_Detection.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

cv::Mat preprocessImage(const cv::Mat& image) {
    cv::Mat denoisedImage;
    cv::fastNlMeansDenoisingColored(image, denoisedImage, 10, 10, 7, 21);
    cv::Mat enhancedImage;
    cv::convertScaleAbs(denoisedImage, enhancedImage, 1, 30);
    cv::Mat cannyImage;
    cv::Canny(enhancedImage, cannyImage, 50, 150);
    return cannyImage;
}

std::vector<std::vector<cv::Point>> detectQuadrilaterals(const cv::Mat& image) {
    std::vector<std::vector<cv::Point>> contours;
    std::vector<cv::Vec4i> hierarchy;
    cv::findContours(image, contours, hierarchy, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);
    std::vector<std::vector<cv::Point>> quadrilaterals;
    for (const auto& contour : contours) {
        double area = cv::contourArea(contour);
        if (area > 2000) {
            std::vector<cv::Point> approx;
            cv::approxPolyDP(contour, approx, 0.01 * cv::arcLength(contour, true), true);
            if (approx.size() == 4) {
                quadrilaterals.push_back(approx);
            }
        }
    }
    return quadrilaterals;
}

cv::Mat drawPoints(cv::Mat image, const std::vector<std::vector<cv::Point>>& quadrilaterals) {
    cv::Mat annotatedImage = image.clone();
    for (const auto& quad : quadrilaterals) {
        cv::drawContours(annotatedImage, quad, -1, cv::Scalar(255, 0, 0), 2);
        for (int i = 0; i < quad.size(); ++i) {
            cv::circle(annotatedImage, quad[i], 5, cv::Scalar(0, 0, 255), -1);
            std::cout << "角点" << i + 1 << " 坐标: (" << quad[i].x << ", " << quad[i].y << ")" << std::endl;
        }
        cv::Moments M = cv::moments(quad);
        if (M.m00 != 0) {
            int cx = static_cast<int>(M.m10 / M.m00);
            int cy = static_cast<int>(M.m01 / M.m00);
            cv::circle(annotatedImage, cv::Point(cx, cy), 5, cv::Scalar(0, 0, 255), -1);
            std::cout << "中点 坐标: (" << cx << ", " << cy << ")" << std::endl;
        }
    }
    return annotatedImage;
}

int main(int argc, char** argv) {
    cv::Mat image = cv::imread("week_9/src/img_1.jpeg");
    if (image.empty()) {
        std::cerr << "无法读取图像文件!" << std::endl;
        return -1;
    }

    cv::Mat processedImage = preprocessImage(image);
    std::vector<std::vector<cv::Point>> quadrilaterals = detectQuadrilaterals(processedImage);
    cv::Mat annotatedImage = drawPoints(image, quadrilaterals);

    cv::namedWindow("Annotated Image", cv::WINDOW_NORMAL);
    cv::imshow("Annotated Image", annotatedImage);
    cv::waitKey(0);
    cv::destroyAllWindows();

    return 0;
}